hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/auth/IAMInstanceCredentialsProvider.java

```diff
@@ -78,23 +117,52 @@ public AwsCredentials resolveCredentials() {
    *
    * @return credentials
    */
-  private AwsCredentials getCredentials() {
+  private synchronized AwsCredentials getCredentials() {
```

Comment:
shouldn't this have been synchronized from the start?

Reply:
no, because it was creating and collecting credentials on every call. so if multiple requests came in, it instantiated new ones and did new HTTP calls. inefficient and clearl a bit brittle. now we want to only have one instance which we close() after